apiserver/facades/client/application/application.go

```diff
@@ -457,14 +457,24 @@ func (api *APIBase) Deploy(args params.ApplicationsDeploy) (params.ErrorResults,
 // the params.ApplicationsDeploy changed. Thus it appears you can use
 // a juju 3.x client to deploy from a juju 2.9 controller, however it
 // fails because the series was not found. Make those corrections here.
+//
+// However, we still need to support some clients which do not pass OS
+// data in the origin, only the param series/charmurl. If a series is found,
+// in either the origin or arg.Series, return early. origin.Series takes
+// priority
 func compatibilityApplicationDeployArgs(arg params.ApplicationDeploy) (params.ApplicationDeploy, error) {
 	origin := arg.CharmOrigin
 	if origin.Series != "" {
+		arg.Series = origin.Series
+		return arg, nil
+	}
```

Comment:
If both `origin.Series` and `arg.Series` are not "", then this will pick the one in the origin, overwriting the series in the arg. There might not be any problems with that, but it might worth to make that nuance explicit by either a comment line or in a unit test. Just wanted to point it out, especially for the case where those two differ from each other.

Reply:
I mention this very briefly in the comment above 'origin.Series take priority'. I picked this because that is the minimal change in behaviour.

Some more eyes on this would be useful though